From: Lennart Poettering <lennart@poettering.net>
Subject: gpt-auto-generator: do not apply image policy on the root fs and /usr/ fs

At the moment the gpt-auto generator does its things we already
transitioned into the host OS, i.e. the root fs and /usr/ are mounted.
Hence suppress image policy checks for those two partitions.

This actually matters, because the root hash/usr hash is taken into
consideration for the image policy checks, but we don't have that in
gpt-auto and hence would refuse operation claiming policy conflicts
event though we never actually operate on the root fs via the dissection
logic.

--- a/src/shared/image-policy.h
+++ b/src/shared/image-policy.h
@@ -94,6 +94,8 @@
 bool image_policy_equal(const ImagePolicy *a, const ImagePolicy *b);       /* checks if defined the same way, i.e. has literally the same ruleset */
 int image_policy_equivalent(const ImagePolicy *a, const ImagePolicy *b);   /* checks if the outcome is the same, i.e. for all partitions results in the same decisions. */
 
+int image_policy_ignore_designators(const ImagePolicy *policy, const PartitionDesignator *designators, size_t n_designators, ImagePolicy **ret);
+
 static inline ImagePolicy* image_policy_free(ImagePolicy *p) {
         return mfree(p);
 }

--- a/src/shared/image-policy.c
+++ b/src/shared/image-policy.c
@@ -613,6 +613,50 @@
         return true;
 }
 
+int image_policy_ignore_designators(
+                const ImagePolicy *policy,
+                const PartitionDesignator *designators,
+                size_t n_designators,
+                ImagePolicy **ret) {
+
+        _cleanup_(image_policy_freep) ImagePolicy *np = NULL;
+        size_t n = 0;
+
+        assert(policy);
+        assert(designators || n_designators == 0);
+        assert(ret);
+
+        np = image_policy_new(_PARTITION_DESIGNATOR_MAX);
+        if (!np)
+                return -ENOMEM;
+
+        np->default_flags = policy->default_flags;
+
+        for (PartitionDesignator d = 0; d < _PARTITION_DESIGNATOR_MAX; d++) {
+                bool ignore = false;
+                for (size_t i = 0; i < n_designators; i++)
+                        if (designators[i] == d) {
+                                ignore = true;
+                                break;
+                        }
+
+                PartitionPolicyFlags flags = ignore ? PARTITION_POLICY_IGNORE : image_policy_get_exhaustively(policy, d);
+                if (flags < 0)
+                        return flags;
+                if (flags == np->default_flags)
+                        continue;
+
+                np->policies[n++] = (PartitionPolicy) {
+                        .designator = d,
+                        .flags = flags,
+                };
+        }
+
+        np->n_policies = n;
+        *ret = TAKE_PTR(np);
+        return 0;
+}
+
 int config_parse_image_policy(
                 const char *unit,
                 const char *filename,

--- a/src/gpt-auto-generator/gpt-auto-generator.c
+++ b/src/gpt-auto-generator/gpt-auto-generator.c
@@ -826,6 +826,15 @@
         _cleanup_free_ char *devname = NULL;
         int r;
 
+        static const PartitionDesignator ignore_designators[] = {
+                PARTITION_ROOT,
+                PARTITION_ROOT_VERITY,
+                PARTITION_ROOT_VERITY_SIG,
+                PARTITION_USR,
+                PARTITION_USR_VERITY,
+                PARTITION_USR_VERITY_SIG,
+        };
+
         assert(!in_initrd());
 
         /* Run on the final root fs (not in the initrd), to mount auxiliary partitions, and hook in rw
@@ -841,6 +850,14 @@
                 return log_debug_errno(r, "Failed to get device node of " DEVNUM_FORMAT_STR ": %m",
                                        DEVNUM_FORMAT_VAL(devnum));
 
+        _cleanup_(image_policy_freep) ImagePolicy *image_policy = NULL;
+        r = image_policy_ignore_designators(
+                        arg_image_policy ?: &image_policy_host,
+                        ignore_designators, ELEMENTSOF(ignore_designators),
+                        &image_policy);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to mark root/usr designators as ignore in image policy: %m");
+
         /* Let's take a LOCK_SH lock on the block device, in case udevd is already running. If we don't take
          * the lock, udevd might end up issuing BLKRRPART in the middle, and we don't want that, since that
          * might remove all partitions while we are operating on them. */
@@ -852,7 +869,7 @@
                         loop,
                         /* verity= */ NULL,
                         /* mount_options= */ NULL,
-                        arg_image_policy ?: &image_policy_host,
+                        image_policy,
                         DISSECT_IMAGE_GPT_ONLY|
                         DISSECT_IMAGE_USR_NO_ROOT|
                         DISSECT_IMAGE_DISKSEQ_DEVNODE|
